Sources/Sentry/SentryScope.m

```diff
@@ -108,68 +126,98 @@ - (void)addBreadcrumb:(SentryBreadcrumb *)crumb
 {
     [SentryLog logWithMessage:[NSString stringWithFormat:@"Add breadcrumb: %@", crumb]
                      andLevel:kSentryLogLevelDebug];
-    @synchronized(self) {
-        [self.breadcrumbArray addObject:crumb];
-        if ([self.breadcrumbArray count] > self.maxBreadcrumbs) {
-            [self.breadcrumbArray removeObjectAtIndex:0];
+    @synchronized(_breadcrumbArray) {
+        [_breadcrumbArray addObject:crumb];
+        if ([_breadcrumbArray count] > self.maxBreadcrumbs) {
+            [_breadcrumbArray removeObjectAtIndex:0];
         }
     }
     [self notifyListeners];
 }
 
 - (void)clear
 {
-    @synchronized(self) {
-        self.breadcrumbArray = [NSMutableArray new];
-        self.userObject = nil;
-        self.tagDictionary = [NSMutableDictionary new];
-        self.extraDictionary = [NSMutableDictionary new];
-        self.contextDictionary = [NSMutableDictionary new];
-        self.distString = nil;
-        self.environmentString = nil;
-        self.levelEnum = kSentryLevelNone;
-        self.fingerprintArray = [NSMutableArray new];
-        self.attachmentArray = [NSMutableArray new];
+    // As we need to synchronize the accesses of the arrays and dictionaries and we use the
+    // references instead of self we remove all objects instead of creating new instances. Removing
+    // all objects is usually O(n). This is acceptable as we don't expect a huge amount of elements
+    // in the arrays or dictionaries, that would slow down the performance.
+    @synchronized(_breadcrumbArray) {
+        [_breadcrumbArray removeAllObjects];
+    }
+    @synchronized(_tagDictionary) {
+        [_tagDictionary removeAllObjects];
+    }
+    @synchronized(_extraDictionary) {
+        [_extraDictionary removeAllObjects];
+    }
+    @synchronized(_contextDictionary) {
+        [_contextDictionary removeAllObjects];
+    }
+    @synchronized(_fingerprintArray) {
+        [_fingerprintArray removeAllObjects];
     }
+    @synchronized(_attachmentArray) {
+        [_attachmentArray removeAllObjects];
+    }
+
+    self.userObject = nil;
+    self.distString = nil;
+    self.environmentString = nil;
+    self.levelEnum = kSentryLevelNone;
+
     [self notifyListeners];
 }
 
 - (void)clearBreadcrumbs
 {
-    @synchronized(self) {
-        [self.breadcrumbArray removeAllObjects];
+    @synchronized(_breadcrumbArray) {
+        [_breadcrumbArray removeAllObjects];
     }
     [self notifyListeners];
 }
 
+- (NSArray<SentryBreadcrumb *> *)breadcrumbs
+{
+    @synchronized(_breadcrumbArray) {
+        return _breadcrumbArray.copy;
```

Comment:
Now that only a copy is returned things should be safe, with the trade off of a copy on each property accessor.

This is a shallow copy though, right?
It's important to point out that if the map or list contain items that are not immutable, data changed in one scope would affect another (i.e: context.device.memory = x) since the same device reference exists in all scope objects. Also, that means we now have `device` or whatever object accessed concurrently. But we should already have anything that exists in `Scope` to be thread safe.

Reply:
Yes, we have the tradeoff of copying on each accessor, but we only use these properties internally. We would need to copy them anyway as I don't want the breadcrumbs array to be the same reference after you serialized the scope.

It is a shallow copy, right.